code/datums/components/uplink.dm

```diff
@@ -164,23 +164,30 @@
 			var/datum/uplink_item/I = uplink_items[category][item]
 			if(I.limited_stock == 0)
 				continue
-			if(I.restricted_roles.len && I.discounted == FALSE)
+			if(I.restricted_roles.len && !I.discounted && ishuman(user))
 				var/is_inaccessible = TRUE
-				for(var/R in I.restricted_roles)
-					if(R == user.mind.assigned_role || debug)
-						is_inaccessible = FALSE
+				var/mob/living/carbon/human/H = user
+				if(locate(H.mind.assigned_role) in I.restricted_roles || debug)
+					is_inaccessible = FALSE
+					break
+				if(is_inaccessible)
+					continue
+			if(I.restricted_species && !I.discounted && ishuman(user))
+				var/is_inaccessible = TRUE
+				var/mob/living/carbon/human/H = user
+				if(locate(H.dna.species.id) in I.restricted_species || debug)
+					is_inaccessible = FALSE
+					break
+				if(is_inaccessible)
+					continue
+			if(I.restricted_roundstart_traits && !I.discounted && ishuman(user))
+				var/is_inaccessible = TRUE
+				var/mob/living/carbon/human/H = user
+				if(locate(H.roundstart_quirks) in I.restricted_roundstart_traits || debug)
+					is_inaccessible = FALSE
+					break
```

Comment:
Ahhh, I see what you've done. `roundstart_quirks` is a list, locating a list in a list isn't going to work, not for this anyway

Reply:
```suggestion
				for(var/datum/quirk/Q in H.roundstart_quirks)
					if(locate(Q.type) in I.restricted_roundstart_traits)
						is_inaccessible = FALSE
						break
				if(debug)
					is_inaccessible = FALSE
					break
```